src/Streamly/Internal/Data/Unfold/Source.hs

```diff
@@ -0,0 +1,88 @@
+-- |
+-- Module      : Streamly.Internal.Data.Unfold.Source
+-- Copyright   : (c) 2021 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- A stream source which can be unfolded with some possibly buffered data.
+-- Allows to 'unread' data i.e. push unused data back to the source. This is
+-- useful in parsing applications with backtracking.
+--
+
+module Streamly.Internal.Data.Unfold.Source
+    ( Source
+
+    -- * Creation
+    , source
+
+    -- * Transformation
+    , unread
+
+    -- * Consumption
+    , isEmpty
+    , read
+    )
+where
+
+#include "inline.hs"
+
+import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
+import Streamly.Internal.Data.Unfold.Resume.Type (Unfold(..))
+import Prelude hiding (read)
+
+-- | An unfold seed with some buffered data. It allows us to 'unread' or return
+-- some data after reading it. Useful in backtracked parsing.
+--
+data Source a b = Source [b] (Maybe a)
+
+-- | Make a source from a seed value. The buffer would start as empty. You can
+-- use 'unread' to add to the buffer.
+--
+-- /Internal/
+source :: Maybe a -> Source a b
+source = Source []
+
+
+-- | Return some unused data back to the source. The data is prepended (or
+-- consed) to the source.
+--
+-- /Internal/
+unread :: [b] -> Source a b -> Source a b
+unread xs (Source ys seed) = Source (xs ++ ys) seed
+isEmpty :: Source a b -> Bool
```

Comment:
newline and some documentation before `isEmpty`

Reply:
done.